### PR TITLE
Do not mutate arrays or ReactElements during React reconciliation

### PR DESCRIPTION
**Pull request by trueadm**
Release notes: none

This fixes an issue that we were experiencing, that turned out to be the correct fix after looking at https://github.com/facebook/prepack/pull/1997. We shouldn't mutate arrays in the reconciliation process, instead we should create new ones when we have new values. Also, we should ensure that we don't mutate ReactElements during reconciliation too, instead we should just create new ones. Mutating existing objects and values breaks down completely when we enter nested effects that need to re-evaluate the same original values, so this fixes a bunch of internal issues we were seeing on our internal FB bundle.

### CODE REVIEW

File: src/react/utils.js [hunk 1]
```diff
@@ -235,6 +239,33 @@ export function forEachArrayValue(realm: Realm, array: ObjectValue, mapFunc: Fun
   }
 }
 
+export function mapArrayValue(
+  realm: Realm,
+  array: ArrayValue,
+  mapFunc: (element: Value, descriptor: Descriptor) => Value
+): ArrayValue {
+  let lengthValue = Get(realm, array, "length");
+  invariant(lengthValue instanceof NumberValue, "Invalid length on ArrayValue during reconcilation");
```
Comment (NTillmann) [inline src/react/utils.js]: It really cannot be an `AbstractValue`?

File: src/react/utils.js [hunk 2]
```diff
@@ -235,6 +239,33 @@ export function forEachArrayValue(realm: Realm, array: ObjectValue, mapFunc: Fun
...
+  let length = lengthValue.value;
+  let newArray = Create.ArrayCreate(realm, length);
+  let returnTheNewArray = false;
+
+  for (let i = 0; i < length; i++) {
+    let elementProperty = array.properties.get("" + i);
+    let elementPropertyDescriptor = elementProperty && elementProperty.descriptor;
+    invariant(elementPropertyDescriptor, `Invalid ArrayValue[${i}] descriptor`);
```
Comment (NTillmann) [inline src/react/utils.js]: Array really cannot have holes?

File: src/react/utils.js [hunk 3]
```diff
@@ -235,6 +239,33 @@ export function forEachArrayValue(realm: Realm, array: ObjectValue, mapFunc: Fun
...
+    let elementValue = elementPropertyDescriptor.value;
+    if (elementValue instanceof Value) {
```
Comment (NTillmann) [inline src/react/utils.js]: Is it okay to just silently drop elements with getters/setters?

File: src/react/utils.js (hunk 1 above)
Comment (trueadm) [reply src/react/utils.js]: Nope, it shouldn't be.

File: src/react/utils.js (hunk 2 above)
Comment (trueadm) [reply src/react/utils.js]: Not inside a ReactElement, no.

File: src/react/utils.js (hunk 3 above)
Comment (trueadm) [reply src/react/utils.js]: Yes, these are ReactElement children – it's created via an internal part of React.

**Comment by trueadm** [comment]
@NTillmann The regression tests are the existing tests in this case – if they all pass, then this is all good. The internal case on our bundle that fails is far too complex to create a repro of :/

File: src/react/utils.js (hunk 2 above)
Comment (gaearon) [reply src/react/utils.js]: Can you explain why in more detail? It’s not obvious to me.

Comment (trueadm) [reply src/react/utils.js]: You can't in a ReactElement:

`React.createElement("div", null, , , , "div");` is not valid JS
`return React.createElement("div", null, undefined, undefined, undefined, "div");` isn't holey.

File: src/react/utils.js (hunk 1 above)
Comment (trueadm) [reply src/react/utils.js]: I've made the invariant a TODO incase this ever comes up. It shouldn't though but at least the invariant makes more sense now.

File: src/react/utils.js [hunk 4]
```diff
@@ -220,19 +220,54 @@ export function getUniqueReactElementKey(index?: string, usedReactElementKeys: S
...
+      if (elementValue instanceof Value) {
+        let newElement = mapFunc(elementValue, elementPropertyDescriptor);
+        if (newElement !== elementValue) {
+          returnTheNewArray = true;
+        }
+        Create.CreateDataPropertyOrThrow(realm, newArray, "" + i, newElement);
+        continue;
+      }
     }
+    Create.CreateDataPropertyOrThrow(realm, newArray, "" + i, realm.intrinsics.undefined);
```
Comment (NTillmann) [inline src/react/utils.js]: This turns missing and accessor elements into undefined elements. Is that really intended?

Comment (trueadm) [reply src/react/utils.js]: Yep, that's what I intended to do. We're resolving empty slots into undefined.

**Pull request closed without merging**